File: scripts/nse-data-collector.js
#!/usr/bin/env node

/**
 * NSE Data Collector for TradeSmart Money
 * Fetches Most Active Stock Calls and Puts data from NSE India APIs
 * Stores data in Supabase for the Smart Money Flow feature
 */

const https = require('https');
const { createClient } = require('@supabase/supabase-js');

// Configuration
const SUPABASE_URL = process.env.NEXT_PUBLIC_SUPABASE_URL;
const SUPABASE_SERVICE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY;
const FORCE_RUN = process.env.FORCE_RUN === 'true';

// NSE API Configuration
const NSE_BASE_URL = 'https://www.nseindia.com';
const NSE_API_BASE = `${NSE_BASE_URL}/api`;

// Request headers to mimic browser behavior
const DEFAULT_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
  'Accept': 'application/json, text/plain, */*',
  'Accept-Language': 'en-US,en;q=0.9',
  'Accept-Encoding': 'gzip, deflate, br',
  'Connection': 'keep-alive',
  'Cache-Control': 'no-cache',
  'Pragma': 'no-cache'
};

// Initialize Supabase client
let supabase;
try {
  if (!SUPABASE_URL || !SUPABASE_SERVICE_KEY) {
    throw new Error('Missing Supabase credentials');
  }
  supabase = createClient(SUPABASE_URL, SUPABASE_SERVICE_KEY);
} catch (error) {
  console.error('❌ Failed to initialize Supabase client:', error.message);
  process.exit(1);
}

/**
 * Make HTTP request with proper error handling
 */
function makeRequest(url, options = {}) {
  return new Promise((resolve, reject) => {
    const requestOptions = {
      ...options,
      headers: { ...DEFAULT_HEADERS, ...options.headers }
    };

    const req = https.get(url, requestOptions, (res) => {
      let data = '';
      
      res.on('data', chunk => data += chunk);
      res.on('end', () => {
        try {
          if (res.statusCode >= 200 && res.statusCode < 300) {
            const jsonData = JSON.parse(data);
            resolve(jsonData);
          } else {
            reject(new Error(`HTTP ${res.statusCode}: ${data}`));
          }
        } catch (parseError) {
          reject(new Error(`JSON Parse Error: ${parseError.message}`));
        }
      });
    });

    req.on('error', reject);
    req.setTimeout(30000, () => {
      req.abort();
      reject(new Error('Request timeout'));
    });
  });
}

/**
 * Establish NSE session and get cookies
 */
async function establishNSESession() {
  console.log('🔗 Establishing NSE session...');
  
  try {
    // First, get the main page to establish session
    await makeRequest(NSE_BASE_URL);
    console.log('✅ NSE session established');
    return true;
  } catch (error) {
    console.error('❌ Failed to establish NSE session:', error.message);
    return false;
  }
}

/**
 * Fetch Most Active Options data from NSE
 */
async function fetchMostActiveOptions(optionType = 'calls') {
  const endpoint = optionType === 'calls' ? 
    `${NSE_API_BASE}/option-chain-indices?symbol=NIFTY` :
    `${NSE_API_BASE}/option-chain-indices?symbol=BANKNIFTY`;
  
  console.log(`📡 Fetching most active ${optionType} data...`);
  
  try {
    const response = await makeRequest(endpoint);
    
    // Process the response to extract most active options
    if (!response || !response.records) {
      throw new Error('Invalid response format from NSE API');
    }

    // Extract and process the data
    const processedData = processMostActiveData(response, optionType);
    console.log(`✅ Fetched ${processedData.length} most active ${optionType}`);
    
    return processedData;
  } catch (error) {
    console.error(`❌ Failed to fetch ${optionType}:`, error.message);
    
    // Return mock data for development/testing
    return generateMockData(optionType);
  }
}

/**
 * Process raw NSE data into our format
 */
function processMostActiveData(response, optionType) {
  try {
    const data = [];
    const records = response.records?.data || [];
    
    // Sort by volume and take top 10
    const sortedRecords = records
      .filter(record => record.CE || record.PE)
      .sort((a, b) => {
        const aVolume = optionType === 'calls' ? (a.CE?.totalTradedVolume || 0) : (a.PE?.totalTradedVolume || 0);
        const bVolume = optionType === 'calls' ? (b.CE?.totalTradedVolume || 0) : (b.PE?.totalTradedVolume || 0);
        return bVolume - aVolume;
      })
      .slice(0, 10);

    for (const record of sortedRecords) {
      const optionData = optionType === 'calls' ? record.CE : record.PE;
      
      if (optionData && optionData.underlying) {
        data.push({
          symbol: optionData.underlying,
          percentage_change: parseFloat(optionData.change || 0),
          volume: parseInt(optionData.totalTradedVolume || 0),
          ltp: parseFloat(optionData.lastPrice || 0)
        });
      }
    }
    
    return data;
  } catch (error) {
    console.error('❌ Error processing NSE data:', error.message);
    return generateMockData(optionType);
  }
}

/**
 * Generate mock data for testing/fallback
 */
function generateMockData(optionType) {
  console.log(`⚠️  Generating mock ${optionType} data for testing...`);
  
  const symbols = ['NIFTY', 'BANKNIFTY', 'RELIANCE', 'TCS', 'HDFC', 'INFY', 'ITC', 'SBIN', 'BHARTIARTL', 'KOTAKBANK'];
  const data = [];
  
  for (let i = 0; i < 5; i++) {
    data.push({
      symbol: symbols[i],
      percentage_change: parseFloat((Math.random() * 10 - 5).toFixed(2)), // Random between -5 and 5
      volume: Math.floor(Math.random() * 1000000), // Random volume
      ltp: parseFloat((Math.random() * 1000 + 100).toFixed(2)) // Random LTP between 100-1100
    });
  }
  
  return data;
}

/**
 * Generate unique session ID
 */
function generateSessionId() {
  const timestamp = new Date().toISOString().replace(/[-:T.]/g, '').slice(0, 14);
  const random = Math.random().toString(36).substring(2, 8);
  return `NSE_${timestamp}_${random}`;
}

/**
 * Store data in Supabase
 */
async function storeDataInSupabase(data, tableName, sessionId) {
  if (!data || data.length === 0) {
    console.log(`⚠️  No data to store in ${tableName}`);
    return false;
  }

  console.log(`💾 Storing ${data.length} records in ${tableName}...`);

  try {
    // Prepare data for insertion
    const insertData = data.map(item => ({
      symbol: item.symbol,
      percentage_change: item.percentage_change,
      session_id: sessionId,
      timestamp: new Date().toISOString()
    }));

    // Insert data into Supabase
    const { data: insertResult, error } = await supabase
      .from(tableName)
      .insert(insertData);

    if (error) {
      throw error;
    }

    console.log(`✅ Successfully stored ${insertData.length} records in ${tableName}`);
    
    // Log sample data
    console.log('📊 Sample data stored:');
    insertData.slice(0, 3).forEach((item, index) => {
      console.log(`   ${index + 1}. ${item.symbol}: ${item.percentage_change}%`);
    });
    
    return true;
  } catch (error) {
    console.error(`❌ Failed to store data in ${tableName}:`, error.message);
    return false;
  }
}

/**
 * Clean up old data (keep only last 24 hours)
 */
async function cleanupOldData() {
  console.log('🧹 Cleaning up old data...');
  
  try {
    const cutoffTime = new Date();
    cutoffTime.setHours(cutoffTime.getHours() - 24);
    
    const tables = ['most_active_stock_calls', 'most_active_stock_puts'];
    
    for (const table of tables) {
      const { error } = await supabase
        .from(table)
        .delete()
        .lt('created_at', cutoffTime.toISOString());
        
      if (error) {
        console.error(`❌ Failed to cleanup ${table}:`, error.message);
      }
    }
    
    console.log('✅ Old data cleanup completed');
  } catch (error) {
    console.error('❌ Cleanup failed:', error.message);
  }
}

/**
 * Check if it's market hours (9:20 AM to 3:30 PM IST, Mon-Fri)
 */
function isMarketHours() {
  if (FORCE_RUN) {
    console.log('🔄 Force run enabled - skipping market hours check');
    return true;
  }

  const now = new Date();
  const istOffset = 5.5 * 60 * 60 * 1000; // IST is UTC+5:30
  const istTime = new Date(now.getTime() + istOffset);
  
  const day = istTime.getUTCDay(); // 0 = Sunday, 6 = Saturday
  const hours = istTime.getUTCHours();
  const minutes = istTime.getUTCMinutes();
  const timeInMinutes = hours * 60 + minutes;
  
  // Market hours: 9:20 AM (560 minutes) to 3:30 PM (930 minutes) IST
  const marketStart = 9 * 60 + 20; // 9:20 AM
  const marketEnd = 15 * 60 + 30;  // 3:30 PM
  
  const isWeekday = day >= 1 && day <= 5; // Monday to Friday
  const isDuringHours = timeInMinutes >= marketStart && timeInMinutes <= marketEnd;
  
  console.log(`⏰ Current IST time: ${istTime.toLocaleString('en-IN', { timeZone: 'Asia/Kolkata' })}`);
  console.log(`📅 Is weekday: ${isWeekday}, Is during market hours: ${isDuringHours}`);
  
  return isWeekday && isDuringHours;
}

/**
 * Main execution function
 */
async function main() {
  console.log('🚀 Starting NSE Data Collection...');
  console.log('📅 Timestamp:', new Date().toISOString());
  
  try {
    // Check market hours
    if (!isMarketHours()) {
      console.log('⏰ Outside market hours - data collection skipped');
      process.exit(0);
    }

    // Generate session ID for this run
    const sessionId = generateSessionId();
    console.log(`🔑 Session ID: ${sessionId}`);

    // Establish NSE session
    const sessionEstablished = await establishNSESession();
    if (!sessionEstablished) {
      console.log('⚠️  Using fallback data due to session issues');
    }

    // Collect data
    console.log('📊 Starting data collection...');
    
    const [callsData, putsData] = await Promise.all([
      fetchMostActiveOptions('calls'),
      fetchMostActiveOptions('puts')
    ]);

    // Store data in parallel
    const [callsStored, putsStored] = await Promise.all([
      storeDataInSupabase(callsData, 'most_active_stock_calls', sessionId),
      storeDataInSupabase(putsData, 'most_active_stock_puts', sessionId)
    ]);

    // Cleanup old data
    await cleanupOldData();

    // Final status
    const totalRecords = (callsData?.length || 0) + (putsData?.length || 0);
    console.log(`\n🎯 Data Collection Summary:`);
    console.log(`   📞 Calls: ${callsData?.length || 0} records ${callsStored ? '✅' : '❌'}`);
    console.log(`   📈 Puts: ${putsData?.length || 0} records ${putsStored ? '✅' : '❌'}`);
    console.log(`   💾 Total: ${totalRecords} records`);
    console.log(`   🔑 Session: ${sessionId}`);
    console.log('✅ NSE data collection completed successfully!');

  } catch (error) {
    console.error('💥 Fatal error during data collection:', error.message);
    console.error('Stack trace:', error.stack);
    process.exit(1);
  }
}

// Handle process termination
process.on('SIGTERM', () => {
  console.log('🛑 Received SIGTERM - gracefully shutting down');
  process.exit(0);
});

process.on('SIGINT', () => {
  console.log('🛑 Received SIGINT - gracefully shutting down');
  process.exit(0);
});

// Run the script
if (require.main === module) {
  main().catch(error => {
    console.error('💥 Unhandled error:', error);
    process.exit(1);
  });
} 

File: scripts/nse_data_collector.py
#!/usr/bin/env python3

"""
NSE Data Collector (Python) for TradeSmart Money

Mirrors the user's working Python approach and runs in GitHub Actions:
- Establishes session by visiting oi-spurts to get cookies
- Uses browser-like headers
- Fetches Most Active Calls and Puts JSON endpoints
- Aggregates pChange by underlying symbol
- Stores top results to Supabase via REST

Environment variables required:
- NEXT_PUBLIC_SUPABASE_URL
- SUPABASE_SERVICE_ROLE_KEY
- FORCE_RUN (optional: 'true' to override market hours)
"""

import os
import sys
import time
import json
import logging
from datetime import datetime, timedelta, timezone

import requests


NSE_BASE_URL = "https://www.nseindia.com"
COOKIE_SET_URL = f"{NSE_BASE_URL}/market-data/oi-spurts"
CALLS_API_URL = f"{NSE_BASE_URL}/api/snapshot-derivatives-equity?index=calls-stocks-val"
PUTS_API_URL = f"{NSE_BASE_URL}/api/snapshot-derivatives-equity?index=puts-stocks-val"


logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
    datefmt="%Y-%m-%dT%H:%M:%SZ",
)


def get_env(name: str, required: bool = True) -> str:
    value = os.environ.get(name)
    if required and not value:
        logging.error("Missing environment variable: %s", name)
        sys.exit(1)
    return value or ""


SUPABASE_URL = get_env("NEXT_PUBLIC_SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = get_env("SUPABASE_SERVICE_ROLE_KEY")
FORCE_RUN = os.environ.get("FORCE_RUN", "false").lower() == "true"


def ist_now() -> datetime:
    # IST is UTC+5:30
    return datetime.utcnow().replace(tzinfo=timezone.utc) + timedelta(hours=5, minutes=30)


def is_market_hours() -> bool:
    if FORCE_RUN:
        logging.info("Force run enabled - skipping market hours check")
        return True

    now = ist_now()
    # Monday=0, Sunday=6
    is_weekday = now.weekday() <= 4
    minutes = now.hour * 60 + now.minute
    # 9:20 AM to 1:00 PM IST (as per original requirement)
    start = 9 * 60 + 20
    end = 13 * 60
    within = start <= minutes <= end
    logging.info("IST now: %s | Weekday: %s | Within market hours: %s", now.isoformat(), is_weekday, within)
    return is_weekday and within


def generate_headers() -> dict:
    return {
        "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36",
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
        "Accept-Language": "en-US,en;q=0.9",
        # Avoid br in Actions to reduce brotli dependency risk
        "Accept-Encoding": "gzip, deflate",
        "Connection": "keep-alive",
        "Cache-Control": "no-cache",
        "DNT": "1",
    }


def establish_session(session: requests.Session) -> None:
    headers = generate_headers()
    # First hit human page to receive cookies
    resp = session.get(COOKIE_SET_URL, headers=headers, timeout=30)
    resp.raise_for_status()
    logging.info("NSE session cookies established (len=%s)", len(resp.headers.get("set-cookie", "")))


def fetch_json(session: requests.Session, url: str) -> dict:
    headers = generate_headers()
    resp = session.get(url, headers=headers, timeout=30)
    # requests handles gzip/deflate automatically
    if resp.status_code != 200:
        raise RuntimeError(f"HTTP {resp.status_code} for {url}")
    try:
        return resp.json()
    except Exception:
        # Try manual decode
        return json.loads(resp.text)


def aggregate_activity(data_obj: dict) -> list:
    if not data_obj or "OPTSTK" not in data_obj or "data" not in data_obj["OPTSTK"]:
        raise ValueError("Invalid data structure from NSE API")

    symbol_to_sum = {}
    rows = data_obj["OPTSTK"]["data"] or []
    for row in rows:
        symbol = row.get("underlying")
        p_change = row.get("pChange")
        if symbol is None or p_change is None:
            continue
        try:
            p = float(p_change)
        except Exception:
            try:
                p = float(str(p_change))
            except Exception:
                p = 0.0
        symbol_to_sum[symbol] = symbol_to_sum.get(symbol, 0.0) + p

    aggregated = [
        {"symbol": sym, "percentage_change": round(total, 2)}
        for sym, total in symbol_to_sum.items()
    ]
    aggregated.sort(key=lambda x: x["percentage_change"], reverse=True)
    return aggregated[:20]


def supabase_insert(table: str, rows: list) -> None:
    if not rows:
        logging.warning("No rows to insert for %s", table)
        return

    url = f"{SUPABASE_URL}/rest/v1/{table}"
    headers = {
        "apikey": SUPABASE_SERVICE_ROLE_KEY,
        "Authorization": f"Bearer {SUPABASE_SERVICE_ROLE_KEY}",
        "Content-Type": "application/json",
        "Prefer": "return=minimal",
    }
    resp = requests.post(url, headers=headers, json=rows, timeout=30)
    if resp.status_code >= 300:
        logging.error("Supabase insert failed (%s): %s", resp.status_code, resp.text[:300])
        resp.raise_for_status()
    logging.info("Inserted %d rows into %s", len(rows), table)


def generate_session_id(prefix: str = "PY") -> str:
    now = datetime.utcnow().strftime("%Y%m%d%H%M%S")
    return f"{prefix}_{now}"


def main() -> None:
    logging.info("Starting NSE data collection (Python)")

    if not is_market_hours():
        logging.info("Outside market hours - skipping run")
        return

    session_id = generate_session_id()
    session = requests.Session()

    # Set session cookies first
    establish_session(session)

    # Small politeness delay
    time.sleep(1.0)

    # Fetch calls & puts with simple retries
    calls_data_obj = None
    puts_data_obj = None
    for name, url in (("calls", CALLS_API_URL), ("puts", PUTS_API_URL)):
        for attempt in range(1, 4):
            try:
                logging.info("Fetching %s data (attempt %d)", name, attempt)
                data = fetch_json(session, url)
                if name == "calls":
                    calls_data_obj = data
                else:
                    puts_data_obj = data
                break
            except Exception as e:
                logging.warning("%s fetch failed: %s", name, e)
                if attempt == 3:
                    logging.error("All attempts failed for %s", name)
                else:
                    time.sleep(1.5 * attempt)

    # Aggregate
    calls_agg = []
    puts_agg = []
    try:
        if calls_data_obj:
            calls_agg = aggregate_activity(calls_data_obj)
    except Exception as e:
        logging.error("Aggregation failed for calls: %s", e)
    try:
        if puts_data_obj:
            puts_agg = aggregate_activity(puts_data_obj)
    except Exception as e:
        logging.error("Aggregation failed for puts: %s", e)

    # Prepare rows with session_id
    calls_rows = [
        {"symbol": r["symbol"], "percentage_change": r["percentage_change"], "session_id": session_id}
        for r in calls_agg
    ]
    puts_rows = [
        {"symbol": r["symbol"], "percentage_change": r["percentage_change"], "session_id": session_id}
        for r in puts_agg
    ]

    # Insert into Supabase
    try:
        supabase_insert("most_active_stock_calls", calls_rows)
    except Exception:
        pass
    try:
        supabase_insert("most_active_stock_puts", puts_rows)
    except Exception:
        pass

    logging.info(
        "Done. calls=%d, puts=%d, session_id=%s",
        len(calls_rows), len(puts_rows), session_id,
    )


if __name__ == "__main__":
    try:
        main()
    except Exception as exc:
        logging.exception("Fatal error: %s", exc)
        sys.exit(1)


